### PR TITLE
package subtabs + doc link + more spacing + better readme

### PR DESCRIPTION
**Pull request by WebFreak001**
readme is now detected from repo content instead of hardcoding README.md
bitbucket api only tested using external debugging of http calls
gitlab not tested at all due to missing api keys
github works fine
external documentation link to dpldocs.info (asked @adamdruppe for this)
planning support for sample file detection here (using repo content)

part of #265, rewritten and optimized a bit

showing dub registry page in large

---

showing dub registry page in small

---

showing dub registry page on mobile

### CODE REVIEW

File: views/view_package.sample.dt [hunk 1]
```diff
@@ -0,0 +1 @@
+p TODO: render sample #{sample}
```
Comment (wilzbach) [inline views/view_package.sample.dt]: Maybe move to a follow-up PR?

File: views/view_package.dt [hunk 2]
```diff
@@ -124,10 +150,21 @@ block body
 			- if( auto pl = "copyright" in versionInfo )
 				li#copyright= pl.get!string
 
+		- if (isMine)
+			//- TODO: make .actions not conditional on isMine, instead have actions which are always there here too (either as a: button.large for actions or p.detail for descriptions)
```
Comment (wilzbach) [inline views/view_package.dt]: Does this actually make sense to display a greyed "Manage this package" button?

File: views/view_package.dt [hunk 3]
```diff
@@ -87,20 +93,40 @@ block body
...
+					- msettings.urlFilter = &urlFilter;
+					!= filterMarkdown(readmeContents, msettings)
+			- else
+				.repositoryReadme
+			- renderedTabContent = true;
+		- else
+			- foreach (i, sample; sampleURLs)
+				- if (activeTab == "sample_"~i.to!string)
+					- renderedTabContent = true;
+					include view_package.sample
```
Comment (wilzbach) [inline views/view_package.dt]: Maybe move this in a separate PR, s.t. documentation button can be rolled out before this feature is complete?

File: views/view_package.dt [hunk 4]
```diff
@@ -87,20 +93,40 @@ block body
...
+				.repositoryReadme.markdown-body
+					- import vibe.textfilter.markdown : MarkdownFlags, MarkdownSettings, filterMarkdown;
+					- scope msettings = new MarkdownSettings;
+					- msettings.flags = MarkdownFlags.backtickCodeBlocks|MarkdownFlags.noInlineHtml|MarkdownFlags.tables;
+					- msettings.headingBaseLevel = 2;
+					- msettings.urlFilter = &urlFilter;
+					!= filterMarkdown(readmeContents, msettings)
+			- else
+				.repositoryReadme
+			- renderedTabContent = true;
```
Comment (wilzbach) [inline views/view_package.dt]: Another tab we could add would be stats e.g. like this:
https://github.com/dlang/dub-registry/issues/160

File: views/view_package.dt (hunk 2 above)
Comment (WebFreak001) [reply views/view_package.dt]: no not a grayed "Manage this package" button, the button should only be visible if logged in and as the author, but the `.actions` div there should be always visible once we have another action

File: views/view_package.dt (hunk 3 above)
Comment (WebFreak001) [reply views/view_package.dt]: if we go through with the review process quickly we can roll this out early too ;)

File: views/view_package.sample.dt (hunk 1 above)
Comment (WebFreak001) [reply views/view_package.sample.dt]: was planned

File: views/view_package.dt (hunk 4 above)
Comment (wilzbach) [reply views/view_package.dt]: Another one could be versions: http://code.dlang.org/packages/vibe-d/versions

**Comment by adamdruppe** [comment]
i like it

**Comment by WebFreak001** [comment]
arsd actually has a readme now after this PR

**Comment by FraMecca** [comment]
I like it a lot

**Comment by p0nce** [comment]
Obviously better

File: source/dubregistry/registry.d [hunk 5]
```diff
@@ -408,8 +409,31 @@ class DubRegistry {
...
+				// check for all other readmes such as README.txt, README.jp.md, etc.
+				readme = files.countUntil!(a => a.type == RepositoryFile.Type.file && a.path.head.name.asUpperCase.startsWith("README"));
+			}
+
+			if (readme != -1) {
+				rep.readFile(reference.sha, files[readme].path, (scope input) {
+					dbver.readme = input.readAllUTF8();
+					string ext = files[readme].path.head.name;
+					// endsWith doesn't like to work with asLowerCase
+					dbver.readmeMarkdown = ext.endsWith(".md") || ext.endsWith(".MD") || ext.endsWith(".Md") || ext.endsWith(".mD");
```
Comment (wilzbach) [inline source/dubregistry/registry.d]: Yeah the problems is that `asLowerCase` doesn't yield a bidirectional range :/
A workaround is this:

```
"foo.mD".retro.asLowerCase.startsWith(".md".retro)
````

https://run.dlang.io/is/PAuvw4

Though it's not very nice either.
Another one:

```
alias lowerCasePred = (a, b) => a.only.sicmp(b.only) == 0;
writeln("foo.mD".endsWith!lowerCasePred(".md"));
```

https://run.dlang.io/is/gdhzFQ

Comment (WebFreak001) [reply source/dubregistry/registry.d]: uh any problem with this? It's just 4 checks

Comment (s-ludwig) [reply source/dubregistry/registry.d]: Or just `ext.extension.icmp(".md") == 0`.

Comment (WebFreak001) [reply source/dubregistry/registry.d]: well probably not icmp though, we don't want combined unicode characters in the filenames

**Comment by s-ludwig** [comment]
Sounds good.

**Comment by s-ludwig** [comment]
Now also deployed.

**Comment by MartinNowak** [comment]
The doc link is visually confusion, it looks like a tab but it's an external link.

I've also argued a bit against centralized and automatically build documentation for two reasons, reliability of the service [¹] and reservations against that being a proper solution for sloppy or lacking documentation [²].

[¹]: https://forum.dlang.org/post/p7d0c2$8h5$1@digitalmars.com
[²]: https://forum.dlang.org/post/p7e79k$13cp$1@digitalmars.com

It might at least encourage people to fix their docs, so let's see how that works out.
We might also want to have overridable doc links, so that people don't have to maintain 2 doc systems.

**Comment by MartinNowak** [comment]
Moving forward with extracting metadata from unstructured repos and readmes seems like a maitenance intensive way forward. I'd suggest to rather support enabling techniques, such as relative links in READMEs, that would also work across github and code.dlang.org, similarly to the documentation link.

File: public/styles/common.css [hunk 6]
```diff
@@ -172,12 +172,88 @@ h3 {
...
+	border-top-left-radius: 4px;
+	border-top-right-radius: 4px;
+	margin-right: 4pt;
+}
+
+.subtabsHeader > .tab.active, .subtabsHeader > .tab:hover, .subtabsHeader > .tab:focus {
+	background-color: #98312A;
+}
+
+.subtabsHeader > .tab.external::after {
```
Comment (MartinNowak) [inline public/styles/common.css]: tab external is quite a contradiction ;)

File: source/dubregistry/registry.d [hunk 7]
```diff
@@ -408,8 +411,31 @@ class DubRegistry {
...
+			// check exactly for readme.me
+			ptrdiff_t readme;
+			readme = files.countUntil!(a => a.type == RepositoryFile.Type.file && a.path.head.name.asUpperCase.equal("README.MD"));
+			if (readme == -1) {
+				// check exactly for readme
+				readme = files.countUntil!(a => a.type == RepositoryFile.Type.file && a.path.head.name.asUpperCase.equal("README"));
+			}
+			if (readme == -1) {
+				// check for all other readmes such as README.txt, README.jp.md, etc.
+				readme = files.countUntil!(a => a.type == RepositoryFile.Type.file && a.path.head.name.asUpperCase.startsWith("README"));
```
Comment (MartinNowak) [inline source/dubregistry/registry.d]: Nice, somewhat problematic that we cannot render them though.

**Comment by WebFreak001** [comment]
relative links to other MD files sounds like a great idea.

Though I don't see how extracting documentation links from the readme is going to be a maintenance effort. Support the basic and most common ways that people link their docs with so we have like 80-90% of repos covered, and the others can set it via a new input text field which will eventually get added.

File: source/dubregistry/registry.d (hunk 7 above)
Comment (WebFreak001) [reply source/dubregistry/registry.d]: well at least support for README, Readme, readme, ReadMe, etc now. i18n is something we might eventually look into, I have seen multiple japanese repositories with different readme names with this README.en.md pattern

**Comment by MartinNowak** [comment]
> relative links to other MD files sounds like a great idea.

Relative links to anything, including source samples, just like on GH.

> Though I don't see how extracting documentation links from the readme is going to be a maintenance effort.

Cause the info needs to be maintained for GH and be extractable for code.dlang.org metadata parser. Basically README.md already allows people to do anything they want with their page, though indeed some guidance on a good format would be nice.
Typically samples are included in README and full file examples could be linked from their, don't see why we should invent our own sample_ convention and lift the info to the code.dlang.org layout.

**Comment by MartinNowak** [comment]
Very nice changes otherwise btw :).

**Comment by MartinNowak** [comment]
Also client-side tabbing and replacing query params `?tab=info` with anchors `#readme` would remain compatible with other html layouts, e.g. scrolling.
